sqlmesh/core/engine_adapter/spark.py

```diff
@@ -40,12 +61,29 @@ def _insert_overwrite_by_condition(
         where: t.Optional[exp.Condition] = None,
         columns_to_types: t.Optional[t.Dict[str, exp.DataType]] = None,
     ) -> None:
-        if isinstance(query_or_df, (pd.DataFrame, PySparkDataFrame)):
-            self._insert_pyspark_df(
-                table_name, self._ensure_pyspark_df(query_or_df), overwrite=True
-            )
+        table = exp.to_table(table_name)
+        df = self.try_get_pandas_df(query_or_df)
+        pyspark_df = self.try_get_pyspark_df(query_or_df)
+        if self._use_spark_session and (df is not None or pyspark_df):
+            if df is not None:
+                pyspark_df = self._ensure_pyspark_df(df)
+            assert pyspark_df
+            self._insert_pyspark_df(table_name, pyspark_df, overwrite=True)
         else:
-            super()._insert_overwrite_by_condition(table_name, query_or_df, where, columns_to_types)
+            if df is not None:
+                query_or_df = next(
```

Comment:
What about the batch size?

Reply:
Without additional complexity, we can't really batch here. We are doing and overwrite so if we batched then we need to make sure that each batch overwrites a complete partition. 

One option would be to remove this and use the parent and do a non-transactional delete/insert append. My thinking though was that if someone had so much data that this is an issue then they should use PySpark DataFrames.

Reply:
doesn't the first one do an overwrite and the others append?

Reply:
The first one might not overwrite all the partitions. If that was the case then you would end up with duplication.

Reply:
maybe let’s chat about this one monday?

Reply:
Addressed in this PR: https://github.com/TobikoData/sqlmesh/pull/890